### ISSUE TITLE
Favorites display on homepage

### ISSUE THREAD

**Issue by petey** [screwdriver-cd/screwdriver#523]
Users should be able to add a set of pipelines to a sharable collection, and display this collection on the main landing page for screwdriver.

**Comment by petey** [comment]
We are still working out the logistics of the UI for this feature.

**Comment by petey** [comment]
The current designs for this is to create collections based on pipeline keywords (#571). 

Collections:
- [ ] crud API for collections {id, userid, name, description, keywords, isDefault} 
- [ ] collections flyout UI panel to create/edit/remove collections

Dashboard:
- [ ] Display pipelines from current default collection
   - [ ] pipeline name
   - [ ] last event status
   - [ ] open/failing PRs
- [ ] Sortable pipelines, by health, last event time, custom?

TODO:
- [ ] Sharing collections with other users
- [ ] Collections of pipeline ids vs. pipeline keywords

![edit collections](https://user-images.githubusercontent.com/78533/27356398-2694102e-55c3-11e7-877f-ec96d074bbf0.png)

**Comment by petey** [comment]
Initial tests with the current API responses requires about 155 requests to fetch data for 46 pipeline. This needs to be remedied by storing more information in events (last build status), or creating a separate API, maybe under collections, that aggregates the necessary data in one call.

**Comment by tkyi** [comment]
# User Research Update

## Summary

After conducting user research/observations about dashboards, @joelseq and I have come to some conclusions.

### Observations

- keywords - not as useful as anticipated
  - they're cumbersome - require making code changes (in the screwdriver.yaml) in order to update them
  - there's too much overhead - they're difficult to keep track of and require the entire team to do
- amount of detail
  - there's usually too much detail displayed on a dashboard for an everyday developer
    - last build or build status of last run is sufficient information for a dashboard
    - if they want more information, they would deep dive on the pipeline page
- use cases
  - dashboards shouldn't really need to be checked everyday to find out if pipelines are healthy or not
    - developers are more concerned with the health of their deployed apps vs pipelines
    - developers should only be concerned with pipeline/build health if something fails; if something fails, it should notify the developer through Slack or email
  - dashboards would be good places for pipeline overviews/subway maps; they would be even more useful/powerful if they had actionable buttons in place
- searching
  - it would be nice if search results returned results sorted by relevancy - so pipelines the developer has contributed to in general/recently
- grouping
  - would be nice to be able to better group favorited pipelines, otherwise developers would have to scroll through way too many Pipelines to find the one they want

### Conclusion

For the first pass at a user dashboard, we think it would be good just to allow users to create and delete `Collections` (groups/folders) and add and remove pipelines into their `Collections`. They should be able to view their `Collections` as well. `Collections` should display pipeline name, last event status, and number of open PRs.

**Comment by joelseq** [comment]
Here's an updated list of stories based on the research findings and new design:

Collections:

- [ ] CRUD API for Collections { id, userid, name, pipelineIDs }
- [ ] Add / remove pipelines from collections
    - [ ] Add pipelines to collections from the search results page
    - [ ] Add pipelines to a collection from the collection's page
    - [ ] Remove pipelines from a collection from the collection's page
- [ ] Collections flyout UI panel to create/edit/remove collections
- [ ] Display pipelines in a collection
    - [ ] Sort collections alphabetically / most recently added

Dashboard:

- [ ] API fetch to get the latest activity from pipelines in the collections
- [ ] News feed UI to display the latest activity from pipelines in the collections

Working on finishing up the hi-fi mockups and getting some more feedback on it. I'll add them here once they're finished.

**Comment by petey** [comment]
I'm not convinced yet that a separate collection page is necessary right now. I believe that the collections flyout should act as "tabs" for the set of pipelines to display on the dashboard.
We should give the ability to remove pipelines from the collections flyout ui, not a separate page.
As a stretch goal, we could add pipelines to a collection from that pipeline's page.

**Comment by tkyi** [comment]
After some discussion, here are the endpoints we think the API should have:

1. `POST /collections` to create a collection (with name & description & pipelineIds) 
2. `PUT /collections/{id}` to update a collection; be it editing name or replacing pipelineIds 
3. `PATCH /collections/{id}/pipelines` to add pipelines - payload would be a body array of pipelines
4. `DELETE /collections/{id}/pipelines` to remove pipelines - payload would be a body array of pipelines
5. `DELETE /collections/{id}` to delete a collection, along with all data

**Comment by petey** [comment]
That's a very restful approach to things, but there are two considerations to make:

1. Is this consistent with any of our other webservice endpoints? We don't have a single endpoint that uses PATCH. We don't have any endpoints that use DELETE except to remove that resource.
2. What will you use to access these endpoints? Will `ember-data` use PATCH or DELETE in this way without heavy customization, or do you plan on using services and write everything from scratch?

**Comment by joelseq** [comment]
After further discussion, we have decided that the following endpoints are what the API will have:

1. `GET /collections` to get a list of user's collections (with pipelineIds for each collection)
2. `GET /collections/{id}` to get a collection (with pipelines populated)
3. `POST /collections` to create a collection (with name & description & pipelineIds)
4. `PUT /collections/{id}` to update a collection; be it editing name or description or replacing pipelineIds. **NOTE**: If you want to add or remove a pipeline you still have to pass in all the other pipelineIds in the payload
5. `DELETE /collections/{id}` to delete a collection

We have decided against going with the `PATCH /collections/{id}/pipelines` and `DELETE /collections/{id}/pipelines` routes as they are inconsistent with our current approach for specifying endpoints and `ember-data` does not use `PATCH` or `DELETE` in this way out of the box.
